### DISCUSSION TITLE
gluetun claims iptables doesn't have NET_ADMIN capability, but pod has it.

### DISCUSSION THREAD

**Discussion by cbc02009** [qdm12/gluetun#1185] [Q&A]
I'm trying to set up gluetun as a sidecar in my flux deployment, but I keep receiving the following error:
```
❯ kl -n vpn downloads-gateway-pod-gateway-79478bf4d8-dlq4z gluetun
========================================
========================================
=============== gluetun ================
========================================
=========== Made with ❤️ by ============
======= https://github.com/qdm12 =======
========================================
========================================

Running version v3.31.1 built on 2022-09-11T20:25:49.370Z (commit 8f04a05)

🔧 Need help? https://github.com/qdm12/gluetun/discussions/new
🐛 Bug? https://github.com/qdm12/gluetun/issues/new
✨ New feature? https://github.com/qdm12/gluetun/issues/new
☕ Discussion? https://github.com/qdm12/gluetun/discussions/new
💻 Email? quentin.mcgaw@gmail.com
💰 Help me? https://www.paypal.me/qmcgaw https://github.com/sponsors/qdm12
2022-10-04T15:25:21Z INFO [routing] default route found: interface eth0, gateway 10.11.2.207 and assigned IP 10.11.2.189
2022-10-04T15:25:21Z INFO [routing] local ethernet link found: vxlan0
2022-10-04T15:25:21Z INFO [routing] local ethernet link found: eth0
2022-10-04T15:25:21Z INFO [routing] local ipnet found: 10.11.2.207/32
2022-10-04T15:25:21Z INFO [routing] local ipnet found: 172.16.1.0/24
2022-10-04T15:25:21Z ERROR for iptables: NET_ADMIN capability is missing: iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
2022-10-04T15:25:21Z INFO Shutdown successful
```
The pod has `NET_ADMIN` capabilities for sure:
```
- env:
        - name: VPN_TYPE
          value: wireguard
        - name: VPN_INTERFACE
          value: wg0
        - name: FIREWALL
          value: "off"
        - name: DOT
          value: "off"
        - name: HEALTH_TARGET_ADDRESS
          value: 1.1.1.1
        - name: HEALTH_VPN_DURATION_INITIAL
          value: 30s
        envFrom:
        - secretRef:
            name: downloads-gateway-vpnconfig
        image: qmcgaw/gluetun:v3.31.1@sha256:dabf4efdadafd6acce67bf89b11396e5c87dd1e25705b36fc9211730c943db47
        imagePullPolicy: IfNotPresent
        name: gluetun
        resources: {}
        securityContext:
          capabilities:
            add:
            - NET_ADMIN
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
```
Which is why I think that the issue is that the wrong iptables is being used. 

Is there a way to specify that gluetun call `xtables-nft-multi` instead of `xtables-legacy-multi` or am I completely off base and the answer is something else?

I'd be happy to provide any additional info, but I'm not sure what would be useful. The configuration for gluetun can be seen here:  https://github.com/cbc02009/k8s-home-ops/blob/17b7ca5c3dc0c9e87167e19f0d26474911b406d4/cluster/manifests/vpn/downloads-gateway/helmrelease.yaml#L53-L71

**Comment by cbc02009** [comment]
Still having this issue. Nothing I try seems to fix it.

**Comment by qdm12** [comment]
The missing `NET_ADMIN` is usually a common user misconfiguration when `permission denied` is encountered. Obviously in your case it's due to something else. I'd guess running the pod as privileged would solve it but that's quite worst than just giving it net_admin capabilities. Since this is rather tighly coupled with your environment/host, try running interactively an `alpine:3.16` container and find a way to make some iptables work. For example on simple docker:

```sh
docker run -it --rm --cap-add=NET_ADMIN alpine:3.16
apk add iptables ip6tables
# Try one of the following lines
iptables -L
ip6tables -L
iptables-nft -L
ip6tables-nft -L
```

If none work and give a permission denied, I can try adding test for `xtables-nft-multi` and `xtables-legacy-multi` within gluetun. Right now only `iptables` + `iptables-nft` and `ip6tables` + `ip6tables-nft` are checked for at:

https://github.com/qdm12/gluetun/blob/192a7a56a3440cbae62f5a9615ea19cc6021b787/internal/firewall/firewall.go#L40

https://github.com/qdm12/gluetun/blob/192a7a56a3440cbae62f5a9615ea19cc6021b787/internal/firewall/ip6tables.go#L18

Would you also know if `xtables-nft-multi` takes care of ipv6? Or is there a corresponding ipv6 program for it?

**Reply by cbc02009** [reply]
Thank you for the help!

I ran the commands in the alpine container like you posted, and these are the outputs:
```
/ # iptables -L
iptables v1.8.8 (legacy): can't initialize iptables table `filter': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
/ # ip6tables -L
ip6tables v1.8.8 (legacy): can't initialize ip6tables table `filter': Permission denied (you must be root)
Perhaps ip6tables or your kernel needs to be upgraded.
/ # iptables-nft -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
/ # ip6tables-nft -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

It looks like iptables-nft works, but gluetun isn't detecting correctly that it needs to be used? I think maybe the xtables thing is a red herring, but I'm not certain. I'm not super knowledgeable about firewall/routing stuff.

Is there an environment variable that can be set to force use of iptables-nft?

**Reply by cbc02009** [reply]
Oh wait, now I remember, `iptables-nft` exists on my host, but it's a link to `xtables-nft-multi`:
```
❯ ll /usr/bin/iptables-nft
Permissions Size User Date Modified Name
lrwxrwxrwx    17 root 25 Jul 10:16  /usr/bin/iptables-nft -> xtables-nft-multi*
```
`ip6tables` also links to the same binary, so it must have the ability to do both ipv4 and ipv6

**Reply by cbc02009** [reply]
@qdm12 Sorry to bug you on this again, just curious if you have any more insight. Should I move this to a bug report?

**Reply by qdm12** [reply]
Sorry to be so sloww answering (and thanks for bumping again);

It turns out you have a "funny" error where you get permission denied for non-nft iptables, but it works for nft iptables. My code would return as soon as a permission denied would be encountered, hence the error you were seeing. This is now fixed in 5a06d8e155e4470419d4dbd15f6d56ec5472657e you can have a go on the latest image :wink:
